src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+
+  actions.forEach((action) => {
+    const previousState =
+      states.length > 0 ? states[states.length - 1] : { ...state };
+    let newState = { ...previousState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
+        break;
+
+      default:
+        break;
+    }
```

Comment:
[CODE STYLE]: It's good practice to use a `switch` statement when dealing with a limited set of conditions, which you have done here. However, make sure to always include a `default` case for error handling, even if it's just a break statement, to ensure that all cases are handled.